### DISCUSSION TITLE
Running magpie with different scenarios

### DISCUSSION THREAD

**Discussion by KeLiu7** [magpiemodel/magpie#444] [Help] [answered]
Dear magpie community,

I would like to run magpie under different scenarios. Should I only change the settings for the two main drivers of the model below?

> cfg$gms$c09_pop_scenario  <- "SSP2"    # def = SSP2
> cfg$gms$c09_gdp_scenario  <- "SSP2"    # def = SSP2

I also noticed that some other parameters also have scenario options. For example:

cfg$gms$c70_feed_scen <- "ssp2"                    
cfg$gms$c60_res_2ndgenBE_dem <- "ssp2" 

 If what I mentioned above is right, do I need to change them as well or keep the default setting (ssp2)? Thanks.

**Comment by abhimishr** [comment]
In case you mean different SSP scenarios, have a look at the `setScenario` function from the `gms` library. 
Manually setting SSP scenarios by hand could be error-prone. A simple example to follow could be seen [here](https://github.com/magpiemodel/magpie/blob/master/scripts/start/forestry.R). 

We also have a `scenario_config.csv` file in the `config` folder. You can find a list of settings to be changed under every SSP scenario there.

**Reply by KeLiu7** [reply]
> In case you mean different SSP scenarios, have a look at the `setScenario` function from the `gms` library. Manually setting SSP scenarios by hand could be error-prone. A simple example to follow could be seen [here](https://github.com/magpiemodel/magpie/blob/master/scripts/start/forestry.R).
> 
> We also have a `scenario_config.csv` file in the `config` folder. You can find a list of settings to be changed under every SSP scenario there.

Thanks Abhimishr. 

Yes, I do mean different SSP scenarios. For example, if I would like to output the food demand under different SSPs. 

Should I only change the settings as below before running the model?

for(ssp in c("SSP1","SSP2","SSP3")) {

  cfg$title <- .title(paste(ssp,"Ref",sep="-"))

  cfg <- setScenario(cfg,c(ssp))

cfg$gms$c15_food_scenario <- paste0(ssp)

  start_run(cfg, codeCheck = TRUE)

}

As `population growth` and `GDP` are the two main drivers of the model, I think it be enough that we just change the SSP scenario for the two variables. Is it right?  Why have to set the SSP scenario for each output variable that we need?

The default scenario in cfg is SSP2 for many variables. Here, if I would like to output food demand under SSP3. I think we should set the scenarios for all variables relevant with food module as SSP3 as well? Is it reasonble that we set food demand under SSP3 but other variables are SSP2 in the model?

Thanks for your patience. Looking forward to your comments. 


![image](https://user-images.githubusercontent.com/72141690/187326442-03d66559-4f02-45f9-80ce-79a8ebb45327.png)

**Comment by abhimishr** [comment, marked as answer]
The code snippet you pasted in principle could work (I did not test it) but `setScenario(cfg,c(ssp))` already takes care of setting the right switch for `cfg$gms$c15_food_scenario` so that line is not needed.

> As population growth and GDP are the two main drivers of the model, I think it be enough that we just change the SSP scenario for the two variables. Is it right?

While this is true, you should always be careful of not setting for example a SSP3 food demand for SSP2 population and gdp development.  

>  Why have to set the SSP scenario for each output variable that we need?

We do not need to, the `scenario_config.csv` file summarizes the places where SSP specific changes need to be made. You can either set them manually (not recommended) or just with `setScenario` like from the code snippet you pasted.

>  Is it reasonble that we set food demand under SSP3 but other variables are SSP2 in the model?

This will not be correct (conceptually at least). In case you want to set up a model run like this, you have to be careful and explicit in describing your scenario setup and model assumptions.

**Reply by KeLiu7** [reply]
Many thanks for your explanation. 

Now I understand: If I would like to run MAgPIE under different SSPs, I just need to use the scripts below. No need to set different SSPs for specific places.  Generally, SSPs should be consistent for all outputs and this could be realized with `setScenario`, right?

for(ssp in c("SSP1","SSP2","SSP3")) {
cfg$title <- .title(paste(ssp,"Ref",sep="-"))
cfg <- setScenario(cfg,c(ssp))
start_run(cfg, codeCheck = TRUE)
}

**Reply by abhimishr** [reply]
Indeed. From the code snippet you posted, I am not sure what `.title` does (maybe it is a function you defined in your script?)

Also for this to work, you would need to call the `gms` library at the beginning of your script and also load the default `cfg` object using `source("config/default.cfg")` before you call your `for` loop.

**Reply by KeLiu7** [reply]
Ah, `.title` refers to the name of different output folders. Here we run the model three times so we need to set the name for them.

Again, many thanks for your help!
